### DISCUSSION TITLE
.env variables are undefined (not accessible) in api routes after project deployed to vercel

### DISCUSSION THREAD

**Discussion by mtin79** [vercel/next.js#15426] [Help]
hi there,
i hope there might be someone who knows next.js deployment architecture well enough to be able to help me with an issue that i can't find a solution for.

i manage deployment variables in **.env file** like this:
```
VARIABLE_NAME=....
OTHER_NAME=....
NEXT_PUBLIC_NAME=...
```

i need to access VARIABLE_NAME 's value in api routes (functions).
everything works when i run "npm run dev" or "npm run build + npm run start".
but as soon as i deploy the project to vercel via "vercel" or "vercel prod" the value of the VARIABLE_NAME is undefined or empty or not there. strangely enough is that the NEXT_PUBLIC_NAME 's value can be used inside the api routes functions.

if i create a new project with next and do the same thing the .env mechanism works also when i deploy to vercel.

any help is much appreciated!

**Comment by rowin1125** [comment]
I have the same issue atm. Did you solve the issue already?

**Reply by mtin79** [reply]
Hi @rowin1125,
i solved this by 
- deleting the existing deployed project from vercel.
- creating a blank new fresh nextjs project  from scratch and migrating the dependencies, configs and app files. in each step i was checking if the env variables (build and runtime) would be accessible in the api route files.
works again but took a one day of work.

meanwhile i think you could maybe just delete the deployed vercel project and deploy the app as a new one. maybe this could work too. 
if so please update the comments here!

best of luck!

**Reply by jensmeindertsma** [reply]
This might be related to a polyfill missing in Next.js 9.5 that got added to 9.5.1. Please check this by updating your dependencies.

**Reply by rowin1125** [reply]
@jensmeindertsma unfortunately I already updated to 9.5.1. Also I dont do anything special with the config. This is the only configs I added: 
```
/* eslint-disable no-param-reassign */
const optimizedImages = require('next-optimized-images');
const withPlugins = require('next-compose-plugins');
const path = require('path');

module.exports = withPlugins([
  [
    optimizedImages,
    {
      imageTrace: {
        color: '#fff',
      },
      webpack(config) {
        config.resolve.alias.images = path.join(__dirname, 'images');
        return config;
      },
    },
  ],
]);
``` 

Any idea where i should look next, really out of options ☹️

**Reply by jensmeindertsma** [reply]
Have you added the environment variables to your Vercel deployment via the web interface? When you deploy, Vercel does **not** automatically pick up your environment variables from a `.env.local` file due to security concerns .

**Reply by rowin1125** [reply]
Yes I did, even copied the same env variables which are in vercel to the development section. I pulled the keys locally with `vercel env pull .env.local` to check if it provides the right keys and it was all correct

**Reply by jensmeindertsma** [reply]
I think this goes beyond my knowledge from this point. I'd suggest talking to Vercel's support team.

**Reply by rowin1125** [reply]
Rebuilding from scratch and redoploying piece by piece to debug, thanks for your support anyway :D

**Reply by lallenfrancisl** [reply]
How this could be solved is by defining it in nextjs.config.js, not sure why I was also searching for it.
```javascript
module.exports = {
  future: {
    webpack5: true,
  },
  async redirects() {
    return [
      {
        source: '/',
        destination: '/login',
        permanent: true,
      },
    ];
  },
  env: {
    NEXT_PUBLIC_API_URI: process.env.NEXT_PUBLIC_API_URI,
    NEXT_PUBLIC_CLIENT_ID: process.env.NEXT_PUBLIC_CLIENT_ID,
    NEXT_PUBLIC_CLIENT_SECRET: process.NEXT_PUBLIC_CLIENT_SECRET,
  },
};
```
The NEXT_PUBLIC_ is also not needed when doing this

**Reply by teemo-dunamu** [reply]
to solve, check .env files in your deployed server project root. if not, you should copy .env files.

**Reply by MefitHp** [reply]
I got here after 2 days of debugging a Keystone/Next js app an this env what the issue. Thank you so much for your help @lallenfrancisl

**Reply by tolubishops0** [reply]
> I think this goes beyond my knowledge from this point. I'd suggest talking to Vercel's support team.

not only vercel, its the same on netlify. do you still recall how you fixed this
